pkg/logql/rangemapper.go

```diff
@@ -0,0 +1,332 @@
+// expression is aggregated by a vector aggregation expression with a label grouping, the downstream expression can be
+// exactly the same as the initial query concatenated by a `sum` operation. If this is the case, overrideDownstream
+// contains the initial query which will be the downstream expression with a split range interval.
+// Example: `sum by (a) (bytes_over_time)`
+// Is mapped to `sum by (a) (sum without downstream<sum by (a) (bytes_over_time)>++downstream<sum by (a) (bytes_over_time)>++...)`
+func (m RangeMapper) mapRangeAggregationExpr(expr *syntax.RangeAggregationExpr, vectorAggrPushdown *syntax.VectorAggregationExpr) syntax.SampleExpr {
+	rangeInterval := getRangeInterval(expr)
+
+	// in case the interval is smaller than the configured split interval,
+	// don't split it.
+	if rangeInterval <= m.splitByInterval {
+		return expr
+	}
+
+	// We cannot execute downstream queries that would potentially produce a huge amount of series
+	// and therefore would very likely fail.
+	if expr.Grouping == nil && hasLabelExtractionStage(expr) {
+		return expr
+	}
+	switch expr.Operation {
+	case syntax.OpRangeTypeBytes, syntax.OpRangeTypeCount, syntax.OpRangeTypeSum:
+		return m.vectorAggrWithRangeDownstreams(expr, vectorAggrPushdown, syntax.OpTypeSum, rangeInterval)
+	case syntax.OpRangeTypeMax:
+		return m.vectorAggrWithRangeDownstreams(expr, vectorAggrPushdown, syntax.OpTypeMax, rangeInterval)
+	case syntax.OpRangeTypeMin:
+		return m.vectorAggrWithRangeDownstreams(expr, vectorAggrPushdown, syntax.OpTypeMin, rangeInterval)
+	case syntax.OpRangeTypeRate:
+		return m.sumOverFullRange(expr, vectorAggrPushdown, syntax.OpRangeTypeCount, rangeInterval)
+	case syntax.OpRangeTypeBytesRate:
+		return m.sumOverFullRange(expr, vectorAggrPushdown, syntax.OpRangeTypeBytes, rangeInterval)
+	default:
+		// this should not be reachable.
+		// If an operation is splittable it should have an optimization listed.
+		level.Warn(util_log.Logger).Log(
+			"msg", "unexpected range aggregation expression",
+			"operation", expr.Operation,
+		)
+		return expr
+	}
+}
+
+// isSplittableByRange returns whether it is possible to optimize the given sample expression
+func isSplittableByRange(expr syntax.SampleExpr) bool {
+	switch e := expr.(type) {
+	case *syntax.VectorAggregationExpr:
+		_, ok := splittableVectorOp[e.Operation]
+		return ok && isSplittableByRange(e.Left)
+	case *syntax.BinOpExpr:
+		return isSplittableByRange(e.SampleExpr) && isSplittableByRange(e.RHS)
+	case *syntax.LabelReplaceExpr:
+		return isSplittableByRange(e.Left)
+	case *syntax.RangeAggregationExpr:
+		_, ok := splittableRangeVectorOp[e.Operation]
+		return ok
+	default:
+		return false
+	}
+}
+
+// clone returns a copy of the given sample expression
+// This is needed whenever we want to modify the existing query tree.
+func clone(expr syntax.SampleExpr) (syntax.SampleExpr, error) {
```

Comment:
Why not `syntax.Clone(expr)`?

Reply:
`syntax.Clone()` does basically the same thing, but with different types: `Expr->Expr`, whereas here I want `SampleExpr->SampleExpr`. This is mainly for convenience.